pkg/server/token.go

```diff
@@ -0,0 +1,107 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"path/filepath"
+
+	"github.com/k3s-io/k3s/pkg/cluster"
+	"github.com/k3s-io/k3s/pkg/daemons/config"
+	"github.com/k3s-io/k3s/pkg/passwd"
+	"github.com/k3s-io/k3s/pkg/util"
+	"github.com/k3s-io/k3s/pkg/version"
+	"github.com/sirupsen/logrus"
+)
+
+type ServerTokenRequest struct {
+	Action   *string `json:"stage,omitempty"`
+	NewToken *string `json:"newToken,omitempty"`
+}
+
+func getServerTokenRequest(req *http.Request) (ServerTokenRequest, error) {
+	b, err := io.ReadAll(req.Body)
+	if err != nil {
+		return ServerTokenRequest{}, err
+	}
+	result := ServerTokenRequest{}
+	err = json.Unmarshal(b, &result)
+	return result, err
+}
+
+func tokenRequestHandler(ctx context.Context, server *config.Control) http.Handler {
+	return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+		if req.TLS == nil || req.Method != http.MethodPut {
+			resp.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		var err error
+		sTokenReq, err := getServerTokenRequest(req)
+		logrus.Debug("Received token request")
+		if err != nil {
+			resp.WriteHeader(http.StatusBadRequest)
+			resp.Write([]byte(err.Error()))
+			return
+		}
+		if *sTokenReq.Action == "rotate" {
```

Comment:
Right now "rotate" is the only possible action, so I'm wondering if it makes sense to check  "actions" when requesting "/token" url. Are there future actions you are thinking? Maybe checking the token?

Reply:
I can remove it now, I wasn't sure if we needed to add more commands (like how secrets-encryption was multistage).